subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,630 @@
+	struct bt_mesh_schedule_register *item = &srv->data.sch_reg[index];
+	struct bt_mesh_scheduler_scheduled_meta *metadata =
+			&srv->data.metadata[index];
+
+	int64_t current_uptime = k_uptime_get();
+	struct tm *current_local = bt_mesh_time_srv_localtime(srv->time_srv,
+			current_uptime);
+
+	if (!set_year(&sched_time, current_local, item)) {
+		return;
+	}
+
+	if (!set_month(&sched_time, current_local, item)) {
+		return;
+	}
+
+	if (!set_day(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_hour(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_minute(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_second(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	int64_t scheduled_uptime = bt_mesh_time_srv_mktime(srv->time_srv,
+			&sched_time);
+	k_delayed_work_submit(&metadata->delayed_work,
+		K_MSEC(scheduled_uptime > current_uptime ?
+				scheduled_uptime - current_uptime : 0));
+	srv->data.status_bitmap ^= 1u << index;
+}
+
+static void scheduled_action_handle(struct k_work *work)
+{
+	struct bt_mesh_scheduler_scheduled_meta *delayed_work =
+			CONTAINER_OF(work,
+				struct bt_mesh_scheduler_scheduled_meta,
+				delayed_work);
+	struct bt_mesh_scheduler_scheduled_meta *metadata =
+			CONTAINER_OF(delayed_work,
+				struct bt_mesh_scheduler_scheduled_meta,
+				delayed_work);
+	metadata->srv->data.status_bitmap ^= 1u << metadata->index;
+
+	if (metadata->srv->data.sch_reg[metadata->index].action > 2) {
+		return;
+	}
+
+	struct bt_mesh_model *next_sched_mod = NULL;
+	uint16_t model_id =
+		metadata->srv->data.sch_reg[metadata->index].action == 2 ?
+		BT_MESH_MODEL_ID_SCENE_SRV : BT_MESH_MODEL_ID_GEN_ONOFF_SRV;
+	struct bt_mesh_model_transition transition = {
+		.time = model_transition_decode(
+		metadata->srv->data.sch_reg[metadata->index].transition_time),
+		.delay = 0
+	};
+	uint16_t scene =
+		metadata->srv->data.sch_reg[metadata->index].scene_number;
+
+	if (model_id == BT_MESH_MODEL_ID_SCENE_SRV && scene == 0) {
+		return;
+	}
+
+	struct bt_mesh_elem *elem = bt_mesh_model_elem(metadata->srv->mod);
+	uint16_t elem_index = elem->addr - bt_mesh_primary_addr();
+
+	do {
+		struct bt_mesh_model *handled_mod =
+				bt_mesh_model_find(elem, model_id);
+
+		if (model_id == BT_MESH_MODEL_ID_SCENE_SRV &&
+				handled_mod != NULL) {
+			struct bt_mesh_scene_srv *srv =
+			(struct bt_mesh_scene_srv *)handled_mod->user_data;
+
+			bt_mesh_scene_srv_set(srv, scene, &transition);
+		}
+
+		if (model_id == BT_MESH_MODEL_ID_GEN_ONOFF_SRV  &&
+				handled_mod != NULL) {
+			struct bt_mesh_onoff_srv *srv =
+			(struct bt_mesh_onoff_srv *)handled_mod->user_data;
+			struct bt_mesh_onoff_set set = {
+				.on_off =
+			metadata->srv->data.sch_reg[metadata->index].action,
+				.transition = &transition
+			};
+			srv->handlers->set(srv, NULL, &set, NULL);
+		}
+
+		elem_index++;
+		elem = bt_mesh_elem_find(bt_mesh_primary_addr() + elem_index);
```

Comment:
```suggestion
		elem = bt_mesh_elem_find(elem->addr + 1);
```

Also, we need to verify that this is a unicast address before passing it to `find()`, in case the current element has address `0x3fff`

Reply:
I think, my case is better since I get the next element unequivocally. It does not require address checking since the primary address and element index related to each other linearly. I'd prefer to leave as is.

Reply:
Elem find will still return a valid elem if you give it a group address, so I think both methods have the same problem

Reply:
Group address never can be the primary one. My case excludes this mistake :)

Reply:
That is true, but `bt_mesh_primary_addr() + elem_index` can be